### ISSUE TITLE
Distinct Gradle Tasks for Multiple Backends

### ISSUE THREAD

**Issue by rwinch** [asciidoctor/asciidoctor-gradle-plugin#157]
It might be nice to have a task named asciidoctor that depends on tasks of asciidoctor<backend-name> so that individual backends can also be built. For example, if the backends html5 and pdf are configured:
- A task named asciidoctorHtml5 would build the html5 backend
- A task named asciidoctorPdf would build the pdf backend
- A task named asciidoctor would depend on asciidoctorHtml5 and asciidoctorPdf

This would allow users to build the HTML5 backend without building the PDF backend. This is nice when doing live editing (the edits will be twice as fast to appear).

**Comment by ysb33r** [comment]
Maybe we can do that as a rule i.e. if `asciidoctor` exists, then `pdfAsciidoctor` or `asciidoctorPdf` (have to decide which one of the two) only builds the `pdf` backend.

**Comment by ysb33r** [comment]
OR (with a bit of cheating)  we can do `gradle asciidoctor --backend pdf`

**Comment by rwinch** [comment]
I guess I am not that worried about implementation, so much as the end user gains. So I'm open to other suggestions.

I would like to note that my original suggestion is attempting to model the change after existing Gradle paradigms. For example, right now Gradle adds

classes task which depends on the tasks compileGroovy, compileJava, processResources

A few observations:

* AFAIK the separation is broken up into tasks. Admittedly I haven't looked into this in any detail though
* The common phrase "compile" is in the front (that is why I was suggesting asciidoctorPdf vs pdfAsciidoctor".

**Comment by ysb33r** [comment]
> I would like to note that my original suggestion is attempting to model the change after existing Gradle paradigms.
Right, I can see which window you looked through :-)
Looking through another window - the conventions for rules one  then `cleanCompileGroovy` will clean whatever `compileGroovy` does, thus a prefix.

Having said that I actually prefer your suggestion, as it says _I want asciidoctor and I want it in PDF_.

(Interestingly enough I just implemented something similar for the gradle-gnumake-plugin).

**Comment by mojavelinux** [comment]
> Having said that I actually prefer your suggestion, as it says I want asciidoctor and I want it in PDF.

+1

I like the name asciidoctorPdf over pdfAsciidoctor.

**Comment by aalmiray** [comment]
I like the idea in principle. As @ysb33r suggests we could use rules for this. What about the following

    asciidoctor {
        // this instructs how many asciidoctorX tasks will be created
        backends = ['html5', 'pdf']
        // common settings to all backends
        options docktype: 'book'
    }

    asciidoctorPdf {
       // override common settings
    }

Following up, there wouldn't be a default `asciidoctor` task, rather an `asciidoctor` extension. The `asciidoctorj` block will merge with this extension. Preconfigured instances of `AsciidoctorTask` will be added to the build based on the common settings. Developers would still have a chance to define custom tasks of type `AsciidoctorTask`; these tasks may inherit and override common settings.

WDYT?

**Comment by mojavelinux** [comment]
This is very interesting because it hides away the extra stanza unless you need it.

I have one question though. Is it possible to have the backend name as a dynamic nested stanza inside the main one?

```
asciidoctor {
  backends = ['html5', 'pdf']

  pdf {
    // override common settings
  }
}
```

You could still dyamically create the `asciidoctorPdf` task and use the config from the nested clause to populate it. Thus, we would just be talking about a more concise way to write it.

**Comment by ysb33r** [comment]
> Following up, there wouldn't be a default asciidoctor task, rather an asciidoctor extension. The asciidoctorj block will merge with this extension. Preconfigured instances of AsciidoctorTask will be added to the build based on the common settings. Developers would still have a chance to define custom tasks of type AsciidoctorTask; these tasks may inherit and override common settings.
> WDYT?

I think we should go one better by having a global configuration (a _extension_ in Gradle-speak). I have described it in #161.

If we then build upon @aalmiray 's suggestion,

1. `asciidoctor` task will take base configuration from `asciidoctorj`. 
1. `backends` settings will restrict rules to only trigger for cases of asciidoctorXXX. i.e. if `pdf` is not a listed backend, then `asciidoctorPdf` will not be a task. (Not sure how to achive this yet, but can have a think),
1. `asciidoctorPdf` is a proxy task. It reads all of it's config from `asciidoctor`, only applies the `pdf` backend. (This might have to happen in the `afterEvaluate` phase).

> I have one question though. Is it possible to have the backend name as a dynamic nested stanza inside the main one?

@mojavelinux Not sure about that. It might be possible. It thus look neat though.

**Comment by sebersole** [comment]
Yes, you can do that using NamedDomainObject support.

**Comment by jnorthr** [comment]
am using  classpath 'org.asciidoctor:asciidoctor-gradle-plugin:1.5.2' gradle plugin to try to generate 
PDF's from my .adoc files; 
*setting backends = ['pdf']* 

this fails with:

_macbook-pro:MrHakiSamples jimnorthrop$ gradle asciidoctor 
Attributes found in options. Existing attributes will be replaced due to assignment. Please use 'attributes' method instead as current behaviour will be removed in future
:asciidoctor
Passing through unknown backend: pdf
:asciidoctor FAILED

FAILURE: Build failed with an exception._


?? Is it possible to produce PDF files yet, pls. ?

**Comment by jnorthr** [comment]
Just found your post about the converter wrapper - so my gradle looks like this & it produces .pdf as expected! Hurrah :-D

apply plugin: 'org.asciidoctor.gradle.asciidoctor'
apply plugin: 'org.asciidoctor.convert' 
buildscript {
    repositories {
        jcenter()
    }
    dependencies {
        classpath 'org.asciidoctor:asciidoctor-gradle-plugin:1.5.2'
        classpath("org.asciidoctor:asciidoctorj-pdf:1.5.0-alpha.6") 
    }

// trial of asciidoctor plugin
asciidoctor { 
    sourceDir = file('src/asciidoc') // needed as plugin actually wants src/doc/asciidoc but plugin info wrong
    outputDir = file("$buildDir/docs")
    backends = ['pdf'] 
    options = [
        doctype: 'book',
        attributes: [
            'source-highlighter': 'coderay',
            toc                 : 'left'
        ]
    ]
}

**Comment by ysb33r** [comment]
@jnorthr You are using an old (deprecated) way of configuring the plugin. Check the README for the new declarative (and more readable way).

**Comment by jnorthr** [comment]
Knew i was right when i said you knew your biscuits 😉 Thanx for the heads-up !

**Comment by pledbrook** [comment]
I don't know if work has started on this, but I think this would be a good opportunity to refactor the plugin based on the new Gradle configuration model. I'll fork the repo and push some changes that highlight what I'm talking about.

**Comment by pledbrook** [comment]
[This commit](https://github.com/pledbrook/asciidoctor-gradle-plugin/commit/e85982beb9341926ce1f9346da302e4bcc12a6c2) contains the skeleton of the new configuration model. It introduces two new concepts: AsciiDoc source sets and AsciiDoc backends. You configure them like this in the build file:

    apply plugin: org.example.AsciidoctorPlugin

    model {
        sourceSets {
            test(AdocSourceSet) {
    //            srcDir = "src/not/funny"
                srcDir = "src/test/asciidoc"
                includes = ["**/index.asciidoc"] as Set
            }
        }
        backends {
            pdf(AdocBackend) {
                outputDir = "pdf"
            }
        }
    }

Interestingly, Gradle 2.11 makes the core Gradle `LanguageSourceSet` available to plugins, so it might be an idea to use that.

Anyway, the main advantage of this approach is that it automatically generates tasks for each combination of source set and backend. This will be listed when the user runs `gradle tasks`.

**Comment by pledbrook** [comment]
This should also fix my issue with vim and continuous build. The vim swap files keep triggering a build when they're modified because (I believe) the Asciidoctor task is declaring a directory as an input rather than a set of source files.

**Comment by jnorthr** [comment]
Thank you for that news. Using this approach would i be able to generate
all three backends in the same declaration: html5, PDF and docbook ? would
this plugin generate all 3 backends by running this task once, or would it
take 3 runs to get this result ?

my current re-gen of the GPars asciidoctor docs here:
http://gparsdocs.de.a9sapp.eu/ requires several outputs so this would be
idea to help in this project. 😉

**Comment by mojavelinux** [comment]
> would this plugin generate all 3 backends by running this task once, or would it take 3 runs to get this result ?

The plugin may (or may not) be able to run once, but Asciidoctor will always need to be executed completely for each backend (aka converter) since the choice of converter affects parsing.

**Comment by jnorthr** [comment]
yes, hence why i have three different gradles, one for each backend plus
one more to generate everything. Not a problem once you know how.

**Comment by pledbrook** [comment]
@jnorthr The plugin would generate three tasks for each source set. With that example code, you would have `generateDocsToHtml5`, `generateDocsToPdf`, and `generateDocsToDocbook` tasks, where the default source set is named "docs".

The example code also adds a `generateAllDocs` task that simply aggregates all the individual ones, so each task would execute separately. However, it would be easy enough to create another Asciidoctor task that is configured to run with all the backends. From what @mojavelinux is saying, I gather that wouldn't be much more efficient, though.

There is a lot of flexibility in the model, and build authors can still create their own instances of the Asciidoctor task if they so wish. Most folks will simply want to specify the parameters of the model, i.e. the sources and the backends.

**Comment by ysb33r** [comment]
> This should also fix my issue with vim and continuous build. The vim swap files keep triggering a build when they're modified because (I believe) the Asciidoctor task is declaring a directory as an input rather than a set of source files.

If you don;t specify anything then it will apply an `include '**'` to the specified source directory . However, it is a `CopySpec` so you can add an `exclude`.

**Comment by ysb33r** [comment]
> Interestingly, Gradle 2.11 makes the core Gradle LanguageSourceSet available to plugins, so it might be an idea to use that.

Once we walk down this road, we break all compatibility with older versions of `Gradle`. It will need to be stated clearly in the docs which version of Gradle will be required. Keeping in mind that the `1.6` release of `asciidoctorj` will require JDK7 as minimum (due to switch to JRuby 9000), it might be a good time, to take care of thie minimum Gradle vesion as well.

I am of the opinion that one should either target Gradle `2.0` as a base version or `2.10`.

**Comment by ysb33r** [comment]
In addition, @pledbrook, I just wanted to say thanks for going down the new model route. I think it will solve the problem described in the OP in a far better way than  the other solutions described earlier.

**Comment by aalmiray** [comment]
I second @ysb33r here, thank you @pledbrook for putting the effort on using the Gradle model. If all model feature used by the new code rely on Gradle 2.10 then I'm fine with upping the ante and mark is as the minimum required version.

**Comment by jnorthr** [comment]
so many tools i use now do require JDK7 as a minimum, so boosting this tool
to jdk7 as well is necessary for compatibility with other stuff. IMHO.

**Comment by gernotstarke** [comment]
@jnorthr well said. JDK7 imho perfectly acceptable.

**Comment by pledbrook** [comment]
The basics of the new model have been there since 2.4 or 2.5, maybe even earlier. So a minimum requirement of 2.10 is probably unnecessary at this point. I'm currently messing around with adding integration tests for the plugin using the Gradle TestKit. That will make a significant refactoring safer.

**Comment by ysb33r** [comment]
TestKit is only practically useful from `2.9` onwards. You will be better off using `GradleTest` instead if you want to use an older version of Gradle.

**Comment by rwinch** [comment]
@pledbrook Thanks for making progress on this issue! I love the updates you are proposing :)

I do have some very minor feedback. I am curious about the task names. Specifically, I find `generateAllDocs` might be confusing in certain contexts. For example, `generateAllDocs` might be misleading if the user had a Javadoc task on the projects as well (i.e. `generateAllDocs` would not generate the Javadoc). I'm wondering if the task names might be a little more specific to Asciidoc (i.e. `generateAllAsciidocs`).

Once again thank you for your efforts!

**Comment by pledbrook** [comment]
@ysb33r I've raised a separate issue to discuss integration/functional tests: #180

**Comment by pledbrook** [comment]
@rwinch That was from a proof of concept for a presentation I did. None of it is set in stone, particularly the task names. I think you're right about the naming, particularly as the documentation could be just one part of a given project build.

**Comment by jnorthr** [comment]
Any idea when this commit will be ready for a test drive Peter ?

**Comment by pledbrook** [comment]
@jnorthr No idea, sorry. I don't want to commit too much time to it, but we'll see. I wanted to make sure that there were integration tests in place first, so those need implementing too.

**Comment by jnorthr** [comment]
Ok Pete. Let me know when u need a tester.
Thx.

**Comment by jnorthr** [comment]
1 question Dan: any way to chg styling so .sidebars look like a box-out on Righthand side of a pg? 
Thx

**Comment by pledbrook** [comment]
It's likely to be a while before it's a good idea to move to the new software model:

* Gradle 2.11 is the first version with official source set support, and it's currently in RC
* I [can't import a Gradle 2.11 project into IntelliJ](https://youtrack.jetbrains.com/issue/IDEA-151010)
* We can still use the new configuration model, but it's a lot more work
* Essential features/fixes may only come in later versions than 2.11 (hard to say at this point)

I still think this is ultimately the right path to take, and I'm less reluctant to require new versions of Gradle than with other tools since it has been so stable with regard to upgrades. I'll keep playing once I can import the project into IntelliJ.

**Comment by jnorthr** [comment]
As Gradle releases become more frequent, it won't be long before the next one arrives, with the new bits needed to make this happen. And it would also b more mature feature-wise. So building on that would be easier as would testing. 

GR8 news though. Thank you 4 allthat work 😉

**Comment by pledbrook** [comment]
Gradle's new software model seems to be in a greater degree of flux than I'd anticipated. Version 2.12 already introduces lots of changes, although I suspect that they are at the source level, not the binary level. Anyway, just to let you know work continues intermittently.

**Comment by jnorthr** [comment]
Ok, super ! let me know if/when you need test help ;-}

**Comment by pledbrook** [comment]
Quick update: I'm still waiting on a Gradle bug to be fixed.

**Comment by aalmiray** [comment]
You're not fooling anyone Peter ;-)

Also, thanks again for looking into this :+1:

**Comment by pledbrook** [comment]
OK, this is going nowhere, sorry. I'll think of an alternative solution that fits with the existing Gradle model.

All we really need to do is:

* Remove the `asciidoctor` task instance
* Introduce an `asciidoctor` extension that is configured in a similar way
* Add source sets to that extension
* Have the plugin generate Asciidoctor tasks based on the source sets and backends configured in the extension
* Create a lifecycle task that depends on all generated Asciidoctor tasks - maybe `docs`?

Unfortunately we can't have an extension and task with the same name, or at least it results in ambiguity over which one you're configuring in an `asciidoctor` block. However, I strongly believe we need to move to a proper extension rather than having a single `asciidoctor` task instance that gets configured with all sorts of extras.

**Comment by aalmiray** [comment]
Indeed. I'm OK with breaking compatibility in version `2.0.0` in order to expose an `asciidoctor` extension instead of  `asciidoctorj`.

**Comment by ysb33r** [comment]
What about modifiying the model DSL

```groovy
apply plugin: org.example.AsciidoctorPlugin

model {
   asciidoctorj {
     // This is global configured and it is pretty easy to reflect an existing 
     // extension into the new model. 
    // Why not just extend the existing asciidoctorj extensio with additional fields?   
   }

   // This is a toplevel componenet that needs to be added, probably similar to components model block
   documentation {
     // Add Asciidoctor model elemenets to documentation
     test(AsciidoctorSpec) {
       attributes ...
       options ...
        // Sources block can then define both Asciidoc sources and arbitry sources that needs copying.
        sources {
          adoc {
            srcDir 'src/test/asciidoc'
            includes '**/index.asciidoc'
          }
          resources {
            srcDir 'src/test/asciidoc'
            include 'images/**', 'css/**'
          }
       }
       // Backends are turned on by (mentioning) configuring them
       backends {
         pdf {
            // Additional config for PDF goes here
         }
         html5 // <-- Just activates it, without any additional options
       }
     } 
   }
}
```
This should be able to generate tasks `asciidoctorPdf` and `asciidoctorHtml5`.

**NOTES**:

* Frm my experience so far with the new model, I don't think it is possible to build a rich-featured Asciidoctor DSL, by only using with managed model elements, it will require some unmanaged elements as well. (Would be happy to be proved wrong).
*. *The new Gradle softtware model is very complex and any implementation should not be underestimated. I think @pledbrook has already discovered it is far from trivial and he has written on and presented on the new model aready. If there is one thing I definitelt took from GradleSummit last week is that even the Gradle core develeopers and taking a step back from the new model to reflect and might possibly even change course in future*.

**Comment by pledbrook** [comment]
@ysb33r I'm suggesting we drop the idea of using the new software model entirely. I don't think we can commit to it until the Gradle development team make it clear what's going to happen with it.

**Comment by jnorthr** [comment]
yes, second that motion, for consistency too.

**Comment by m0j0hn** [comment]
I created this to enable saying "-Pbackends=pdf,xhtml5", "-Pbackends=pdf", "-Pbackends=xhtml5", or nothing, which uses the default backends specified in the list named "BACKENDS".

asciidoctor {
  // Other configuration goes here

  backends = {
    // Enable -Pbackends='pdf,xhtml5' to specify backend(s) to use
    def BACKENDS = ['pdf','xhtml5'] // Default backends we use
    if(project.hasProperty('backends')){
      def backendsList = project.property('backends').split(',')
      backendsList.each { backend ->
        if (! BACKENDS.contains(backend)) {
          throw new InvalidUserDataException("${backend} is not a supported backend.\nUsage: gradle asciidoc -Pbackends=\'${BACKENDS.join(',')}\'")
        }
      }
      return backendsList
    } else {
      return BACKENDS // Default
    }
  }()

}

https://gist.github.com/m0j0hn/b7a03c532add64c3ec1ea36622a66d8e

**Comment by ysb33r** [comment]
In 2.0 AsciidoctorPdf & AsciidoctorEpub are now separate tasks and needs to be enabled via plugins `ort.asciidoctor.jvm.pdf` and `org.asciidoctor.jvm.epub` respectively. 
With this approach we could deal witht he peculiarities of PDF & EPUB much better.

There is not going to be a asciidoctoHtml5 and asciicoctorDocBook. This can still be dealt with via the  generic `asciidoctor` task which is available via the `org.asciidoctor.jvm.convert` plugin. (You can apply more than one plugin in a project).

I think this issue has been dealt with adequately in #249 and I'm closing this issue.
 Feedback (and issues) on 2.0-alpha.2 will be welcome.
